### ISSUE TITLE
Proposal: Allow custom elements to be in any namespace

### ISSUE THREAD

**Issue by hfmanson** [WICG/webcomponents#634]
In the current spec custom elements can only be in the namespace http://www.w3.org/1999/xhtml 
and without using `is=""` derive from `HTMLElement`. It is desirable to have custom elements in other namespaces
* Create custom SVG elements in the SVG namespace which derive from `SVGElement`
* Create elements in your own namespace using XHTML which derive from `Element`

My proposal is to add  a ‘namespace’ property to ElementDefinitionOptions to specify in which namespace the custom element should be placed. I created [an issue](https://github.com/w3c/webcomponents/issues/328) before for Custom Elements v0.
# Advantages
* Custom element name clashes will no longer occur
* If the namespace isn't a standard namespace (e.g. http://www.w3.org/1999/xhtml or http://www.w3.org/2000/svg) the dash requirement for new custom element tag names can be removed. Updated according to [ExE-Boss](https://github.com/WICG/webcomponents/issues/634#issuecomment-482822301)
* Possible to implement polyfills for XML dialects such as XForms, MathML or X3D
# References
"Distributed Extensibility: Finally Done Right?" in [XML Prague 2014 proceedings](http://archive.xmlprague.cz/2014/files/xmlprague-2014-proceedings.pdf)
[Are web components *seriously* not namespaced?](https://lists.w3.org/Archives/Public/public-webapps/2015JanMar/0452.html) on the w3c mailing list
# Implementation
https://github.com/hfmanson/custom-elements/tree/namespace-aware is a fork that implements this feature. A live demo is at http://mansoft.nl/custom-elements/

This polyfill needs to be run on a browser that does not support Custom Elements natively, e.g. Microsoft Edge or Firefox 58.

**Comment by trusktr** [comment]
This is an interesting topic. It's completely worthy to note that someone can opt into making a UI entirely out of SVG, but then they're unfortunately out of luck if they want to use Web Components. It would be great to use Web Components with any type of element (just like we can use React/Angular/Vue with any type of elements).

Worth noting: [Scoped Custom Element Registries](https://github.com/w3c/webcomponents/issues/716) can also solve some of the namespacing problems.

Maybe, with scoped Custom Element registries, there could be a feature to allow for Custom Elements to define which class that elements distributed into a shadow root are upgraded to. No idea how it'd work, but seems like that topic may lend to that area of discussion. ( @justinfagnani )

**Comment by jarek-foksa** [comment]
I second that. I'm making extensive use of v0 custom SVG elements in [Boxy SVG](https://boxy-svg.com/app) to implement on-canvas editing controls. Encapsulated SVG components could be also useful when doing complex charts and diagrams.

**Comment by annevk** [comment]
We discussed this in Tokyo, there isn't much interest from implementers. And SVG generally isn't getting much attention, despite having loads of technical debt. And to really make use of this SVG would have to be able to use shadow roots too (and its layout model have to work with that too).

**Comment by bzbarsky** [comment]
Note that the original shadow DOM proposals did try to support this.  Would be useful context to know why it was dropped.  @dglazkov

**Comment by jarek-foksa** [comment]
@annevk As discussed [here](https://github.com/w3c/svgwg/issues/122), the needed changes to the SVG layout model seem rather trivial. SVG 2 spec is already defining `<use>` elements using Shacow DOM model: https://svgwg.org/svg2-draft/struct.html#UseShadowTree.

@bzbarsky WebKit team did not want SVG support in v1 specs for the sake of simplicity and ease of implementation: https://github.com/w3c/webcomponents/issues/110#issuecomment-180642629. I don't recall them dismissing the idea altogether.

**Comment by bzbarsky** [comment]
@jarek-foksa Right, but what were the specific complications and difficulties associated with it?  "simplicity" and "ease of implementation" suggests those existed, unless they're code for "don't even want to think about it".

**Comment by bzbarsky** [comment]
In particular, I think I kow what the issues were with supporting this in the initial shadow DOM setup (all sorts of complications around ensuring that the proto chain matched the element type created etc), but most if not all of those should be gone in the `[HTMLConstructor]` setup.  I haven't had a chance to think through what issues, if any, remain, hence asking whether someone else already has.

**Comment by jarek-foksa** [comment]
@bzbarsky I'm not sure about the specifics, but I recall that initial implementation of custom SVG elements in Chromium was ridden with bugs and it took them like 2 years to make it fully usable. For example, the styles and IDs in the shadow DOM were not properly scoped, thus leaking to the global scope and causing naming clashes.

**Comment by bzbarsky** [comment]
Just to be clear, custom elements and shadow DOM are orthogonal in terms of both specification and implementation.

The problem you describe has to do with shadow DOM, and is going to be a problem for shadow DOM even if you only allow HTML elements to be shadow hosts, as long as you put SVG elements in the shadow tree.  It seems completely unrelated to custom elements.

**Comment by jarek-foksa** [comment]
@bzbarsky Yeah, I was assuming that custom SVG elements would make sense only with shadow DOM support as mentioned by Anne. Anyway, in Chromium most issues with ID and style scoping were specific to SVG shadow trees, for example: https://bugs.chromium.org/p/chromium/issues/detail?id=424598

**Comment by hayatoito** [comment]
> SVG 2 spec is already defining <use> elements using Shacow DOM model: https://svgwg.org/svg2-draft/struct.html#UseShadowTree.

That would be just a sort of the implementation detail, and we should be careful not to make a decision based on that.

We should ask and clarify the intention why SVG is trying to use Shadow DOM as an implementation of `<use>` there.
I guess that is just the result of spec-ing how Blink is implementing SVG's `<use>`, where there is no other intention.

**Comment by gertcuykens** [comment]
Hmm is this the proposal that would also make `is` work for `class SvgIcon extends SVGElement `?
```js
      import {html, render} from 'https://unpkg.com/lit-html?module'
      import SvgIcon from './SvgIcon.js'
      customElements.define('svg-icon', SvgIcon, {extends: 'svg'})
      render(html`
        <svg is="svg-icon" viewBox="0 0 75 75" icon="add"></svg>
        `, document.body)
```

Uncaught DOMException: Failed to execute 'define' on 'CustomElementRegistry': "svg" is an HTMLUnknownElement

**Comment by ExE-Boss** [comment]
@gertcuykens It should.

@hfmanson
> If the namespace isn't used by the browser for rendering (e.g. http://www.w3.org/1999/xhtml or http://www.w3.org/2000/svg) the dash requirement for new custom element tag names can be removed.

Maybe keep the dash restriction for official w3c namespaces.

This is also because **MathML** elements are all implemented using just the `Element` interface (there’s no `MathMLElement` **WebIDL** interface and rendering is handled via custom `appearance` and `display` values).

**Comment by bkardell** [comment]
@ExE-Boss  note we are interested in improving the idl situation here https://github.com/mathml-refresh/mathml/issues/83

**Comment by justinfagnani** [comment]
I recently noticed that SVG2 specs the behavior of unknown tags to act like `<g>`: https://www.w3.org/TR/SVG2/struct.html#UnknownElement

The problem is no browser implements this. If they did, we would be able to "polyfill" custom SVG elements much like we do with HTML custom elements.

**Comment by trusktr** [comment]
Another question is, if we have custom SVG elements (as awesome as it would be for developers making things with better-organized SVG code), then we export the DOM as HTML and throw it at applications like Adobe Illustrator or Inkscape, what is supposed to happen?

**Comment by akauppi** [comment]
@trusktr I don't think that would be a problem. Either renderers ignore them altogether (as happens today, with unknown SVG elements in browsers), or they treat them as `g` if SVG2 is followed (see @justinfagnani 's message above).

So to me, this seems a non-concern. Any HTML/XML is intended to be expandable, so nothing new there.

**Comment by justinfagnani** [comment]
Is there anything we could do to get movement on allowing SVG custom elements? There are some very large unaddressed use cases here for charting, dynamic illustrations, drawing apps, very graphical UIs like node/edge editors, etc. Any chance SVG is in a better maintained state since 2018?

**Comment by lordanubi** [comment]
Definitely needing this, use tag is great but has its limits (no custom attributes for example)

**Comment by andreafuturi** [comment]
I think this a really important feature for improving UI

**Comment by nileshtrivedi** [comment]
+1 for allowing custom elements to be in the SVG namespace. Will make the development of diagramming apps much easier as framework-agnostic libraries can emerge.

**Comment by bennyschudel** [comment]
Was somehow surprised that this is not implemented. For dynamic UI this would be a great boost.

**Comment by Delnee** [comment]
I am using the Lit library to build web components, and its a shame that large SVG elements cannot be decomposed into custom components, because this is isn't supported.

**Comment by cscheffauer** [comment]
+1

**Comment by mlarabi** [comment]
Kind support for this would be greatly appreciated. WebComponents have some unique advantages over other frameworks that we cannot currently use.
We have an application built 95% on ReactJS. However some components need to be highly interactive with the user (meaning absolutely nothing else should intervene between the user interaction and the update of the DOM besides strict minimal business logic). Instead of using another framework on top of ours we prefer using WebComponents as it covers our needs and help us achieve our goals.

However the lack of support of other namespace (in our cases the SVG one http://www.w3.org/2000/svg) really hurt the use of said technology.
